### ISSUE TITLE
RFC: Lerna Project

### ISSUE THREAD

**Issue by andrestone** [projen/projen#9]
Hey!

How about projen for bootstrapping and maintaining Lernyarn monorepos?

```typescript
import { LernaProject, TypeScriptLibraryProject, NodeProject } from '../lib';

const lernaProject = new LernaProject({
  name: 'my-monorepo-project',
  packages: [
    {
      name: 'ddb-connector',
      location: 'packages/@services',
      project: new TypeScriptLibraryProject({ name: 'ddb-connector', description: 'DynamoDB API / Connector' })
    },
    {
      name: 'service-portal',
      location: 'packages/@services',
      project: new NodeProject({ name: 'service-portal-api', description: 'Service Portal' })
    }],
  workSpaces: {
    packages: [
      "packages/@web/*",
      "packages/@rn/*",
      "packages/@iac/*",
      "packages/@services/*",
    ],
    nohoist: [],
  }
});

lernaProject.addPackage({
  name: 'geo',
  location: 'packages/@services',
  project: new TypeScriptLibraryProject({ name: 'geocoding-service', description: 'Geocoding Service' })
});
lernaProject.addBootstrappedPackage({
  name: 'cloud-infra',
  location: 'packages/@iac',
  beforeCommand: 'mkdir cloud && cd cloud',
  npxCommand: 'npx cdk init --language typescript',
});
lernaProject.addBootstrappedPackage({
  name: 'core',
  location: 'packages/@web',
  npxCommand: 'npx create-react-app web-frontend',
});

lernaProject.synth();
```

**Comment by eladb** [comment]
Yes please! I've been wanting to do that for a while!

Would you like to use [cdk8s](https://github.com/awslabs/cdk8s) as the guinea pig?

**Comment by andrestone** [comment]
Sure, will have a look at the repo.

**Comment by eladb** [comment]
API design is going to be fun here. Would be fun to discuss over a mini-RFC.

Basically, just create the PR and at least write an initial version of the README (you can add some prototyping code or whatever), but think backwards from the README entry for this feature.

**Comment by andrestone** [comment]
Hmm.. I tend to first do 1:1 implementations from the spec and then start abstract away from those constraints.

So the question would be at which level of abstraction we should settle? Should we pick cdk8s as a minimally opinionated implementation mirror an API on top of the configs specifications?

(In addition to the ideas above and now that we have templates, we could expose an `addPackageFromTemplate` method).

**Comment by andrestone** [comment]
After taking a closer look at the cdk8s repo, I'm now a bit confused about what you meant by using it as the guinea pig. I might be missing something but it feels like cdk8s only uses Lerna to run package commands while all other functionality provided by Lerna is replaced by custom scripts / workflows.

So the question is: should I build an API that implements the cdk8s project but uses Lerna as much as possible, or just an API that supports what cdk8s uses from Lerna?

**Comment by eladb** [comment]
All the repo-wide comments in cdk8s are performed by lerna, not just "package" (see the top-level package.json file).

Based on your proposal above, I like the idea of having a centralized .projenrc.js file that basically describes the entire project.

**Comment by andrestone** [comment]
> All the repo-wide comments in cdk8s are performed by lerna, not just "package" (see the top-level package.json file).

I meant to say commands inside packages 🤣 

But ok. Maybe my last comment is not relevant anyway. I'm working on it. I'll open an WIP PR so you can do some course correction whenever possible.

**Comment by JordanSinko** [comment]
@adamelmore - do you need help with this? Our team loves Lerna and use it for pretty much everything within our org. This would be extremely helpful as we also love projen now!

Let me know!

**Comment by adamelmore** [comment]
> @adamelmore - do you need help with this? Our team loves Lerna and use it for pretty much everything within our org. This would be extremely helpful as we also love projen now!
> 
> 
> 
> Let me know!

The PR I put up was a working implementation; just needed a rebase and review! @eladb any appetite for Lerna support? I can re-open if so.

**Comment by andrestone** [comment]
> The PR I put up was a working implementation; just needed a rebase and review! @eladb any appetite for Lerna support? I can re-open if so.

So was [https://github.com/eladb/projen/pull/13](this one) at the time. 

My guess is that this creates new constructs / hierarchies that don't really fit into projen as it was conceived.

**Comment by eladb** [comment]
@adamelmore There is definitely appetite for lerna support! I like the direction in #13 and would be happy to provide feedback if you'd like to pick this up from where you left off.

**Comment by skyrpex** [comment]
There's also the PNPM workspaces. PNPM uses a `pnpm-workspace.yaml` file, see [here](https://pnpm.js.org/en/pnpm-workspace_yaml). AFAIK, the interface is simpler than the Lerna one.

Maybe we could have a generic workspace project? :thinking: 

```js
const projen = require("projen");

const monorepo = new projen.MonorepoProject({
    name: "acme-monorepo",
    // If NPM or YARN, leverage the workspace management to Lerna. Otherwise, use PNPM directly.
    packageManager: projen.NodePackageManager.PNPM,
    packages: [
        new projen.NodeProject({
            name: "@acme/delay",
            outdir: "packages/delay",
        }),
        new projen.NodeProject({
            name: "@acme/sum",
            outdir: "packages/sum",
        }),
    ],
});

monorepo.addPackage(
    new projen.NodeProject({
        name: "@acme/multiply",
        outdir: "packages/multiply",
    })
);

monorepo.synth();
```

There's some details that would need to be discussed. For example:
- The workspaces listed in the monorepo could be generated using the subproject outdirs. No need to type them manually.
- When using workspaces, there's no need to run `npm install/yarn install/pnpm install` for every workspace. Doing that on the root project is enough.

**Comment by dtothefp** [comment]
Hi would love to help with this and definitely have opinions on managing TS monorepos with Lerna / yarn workspaces. Not sure the best thread to comment on to contribute as I see these others
- https://github.com/projen/projen/issues/450
- https://github.com/projen/projen/issues/14

Anyhow, we have hacked together a projen monorepo at my work and definitely has some rough edges. Generally using it at this point to create CDK and Typescript projects.

Our other in our other monorepos for styleguides, web projects, etc. the most invaluable aspect I've found other than lerna / yarn workspaces is [Typescript project references](https://www.typescriptlang.org/docs/handbook/project-references.html) .  Without having dug into projen source code much I'm not completely sure how the TS is being compiled but as monorepos scale inevitably packages become dependent upon one another. And this is where TS project references becomes invaluable for defining package dependencies so TS compilation doesn't blow up without having to manage a complicate order of TS compilation commands. Additionally, a single like `tsc --build ./packages/tsconfig.packages.json` can be run to compile all packages rather than composing multiple `tsc` commands.

Not sure if you've already looked into this or have other ways you are considering compiling TS in monorepos but would be happy to add a section in the RFC if you think this is a viable route.
